platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/generic/MetadataRendererRegistry.java

```diff
@@ -99,22 +124,25 @@ static String trimIfNecessary(String input) {
         }
     }
 
-    public static final class StringMetadataRenderer implements MetadataRenderer<String> {
+    public static final class UnknownTypeRenderer implements MetadataRenderer {
         @Override
-        public Class<String> getMetadataType() {
-            return String.class;
+        public Set<Class<?>> getMetadataTypes() {
+            return Collections.singleton(Object.class);
         }
 
         @Override
         public SimpleHtmlWriter render(Object metadata, SimpleHtmlWriter htmlWriter) throws IOException {
-            return (SimpleHtmlWriter) htmlWriter.characters(MetadataRenderer.trimIfNecessary((String) metadata));
+            return (SimpleHtmlWriter) htmlWriter
+                .startElement("span").attribute("class", "unrenderable")
+                .characters("[unrenderable type]")
+                .endElement();
         }
-    }
+    };
 
-    public static final class NumberMetadataRenderer implements MetadataRenderer<Number> {
+    public static final class BasicRenderer implements MetadataRenderer {
         @Override
-        public Class<Number> getMetadataType() {
-            return Number.class;
+        public Set<Class<?>> getMetadataTypes() {
+            return new HashSet<>(Arrays.asList(String.class, Number.class, Boolean.class));
```

Comment:
We should cache this, and possibly use an `ImmutableSet` here since it'll be much faster and cleaner.

Reply:
Good ideas.